Framework/DataHandling/src/LoadCSNSNexus.cpp

```diff
@@ -0,0 +1,568 @@
+ */
+std::vector<int64_t>
+LoadCSNSNexus::getPixelId(const vector<string> &inputList) {
+  std::vector<int64_t> _tmp;
+  std::vector<int64_t> pixelId;
+  m_file->openGroup(m_entry, "NXentry");
+  m_file->openGroup("instrument", "NXinstrument");
+  std::map<std::string, std::string> entries = m_file->getEntries();
+  for (auto j = inputList.begin(); j != inputList.end(); j++) {
+    auto it = entries.find(*j);
+    if (it != entries.end()) {
+      m_file->openGroup(it->first, it->second);
+      m_file->readData("pixel_id", _tmp);
+      pixelId.insert(pixelId.end(), _tmp.begin(), _tmp.end());
+      m_file->closeGroup();
+    }
+  }
+  sort(pixelId.begin(), pixelId.end());
+  m_file->closeGroup();
+  m_file->closeGroup();
+  return pixelId;
+}
+
+/**
+ * Return time-of-flight
+ * @param[in] typeName :: module or monitor
+ * @return time-of-flight
+ */
+std::vector<uint32_t> LoadCSNSNexus::getTimeBin(string typeName) {
+  std::vector<int> tmp;
+  std::map<std::string, std::string> entries;
+  m_file->openGroup(m_entry, "NXentry");
+  m_file->openGroup("instrument", "NXinstrument");
+  entries = m_file->getEntries();
+  string detName = "";
+
+  for (auto it = entries.begin(); it != entries.end(); ++it) {
+    string name = it->first;
+    if (name.substr(0, 6) == typeName.substr(0, 6)) {
+      m_file->openGroup(name, it->second);
+      m_file->readData("time_of_flight", tmp);
+      m_file->closeGroup();
+      break;
+    }
+  }
+  m_file->closeGroup();
+  m_file->closeGroup();
+  std::vector<uint32_t> timeBin(tmp.begin(), tmp.end());
+  return timeBin;
+}
+
+/**
+ * Get histogram data
+ * @param[in] inputList :: moduleList or monitorList
+ * @return histogram data
+ */
+std::vector<uint32_t>
+LoadCSNSNexus::getHistData(const vector<string> &inputList) {
+  std::vector<int> tmp;
+  std::vector<uint32_t> data;
+
+  m_file->openGroup("csns", "NXentry");
+  m_file->openGroup("histogram_data", "NXcollection");
+  std::map<std::string, std::string> entries = m_file->getEntries();
+  for (auto j = inputList.begin(); j != inputList.end(); ++j) {
+    auto it = entries.find(*j);
+    if (it != entries.end()) {
+      m_file->openGroup(it->first, it->second);
+      m_file->readData("histogram_data", tmp);
+      data.insert(data.end(), tmp.begin(), tmp.end());
+      m_file->closeGroup();
+    }
+  }
+  m_file->closeGroup();
+  m_file->closeGroup();
+  return data;
+}
+
+/**
+ * load histogram data into workspace
+ * @param[in] workspace ::
+ * @param[in] timeOfFlight ::
+ * @param[in] pidNums :: total pixel numbers
+ * @param[in] histData :: histogram data
+ */
+void LoadCSNSNexus::loadHistData(MatrixWorkspace_sptr &workspace,
+                                 std::vector<uint32_t> &timeOfFlight,
+                                 size_t pidNums,
+                                 std::vector<uint32_t> &histData) {
+  size_t timeNums = timeOfFlight.size();
+  MantidVecPtr x, e;
+  MantidVec xA;
+  MantidVec xRef(timeNums);
+
+  int st1 = 0;
```

Comment:
It is good practice to try and use stl algorithms where possible:
```c++
std::copy(timeOfFlight.cbegin(), timeOfFlight.cend(), xRef.begin());
```
can replace this loop.